homeassistant/components/todo/__init__.py

```diff
@@ -162,6 +172,88 @@ async def async_move_todo_item(
         """
         raise NotImplementedError()
 
+    @final
+    @callback
+    def async_subscribe_updates(
+        self,
+        listener: Callable[[list[JsonValueType] | None], None],
+    ) -> CALLBACK_TYPE:
+        """Subscribe to To-do list item updates.
+
+        Called by websocket API.
+        """
+        if self._update_listeners is None:
+            self._update_listeners = []
+        self._update_listeners.append(listener)
+
+        @callback
+        def unsubscribe() -> None:
+            if self._update_listeners:
+                self._update_listeners.remove(listener)
+
+        return unsubscribe
+
+    @final
+    def async_update_listeners(self) -> None:
+        """Push updated To-do items to all listeners."""
+        if not self._update_listeners:
+            return
+
+        todo_items: list[JsonValueType] = [
+            dataclasses.asdict(item) for item in self.todo_items or ()
+        ]
+        for listener in self._update_listeners:
+            listener(todo_items)
+
+    @callback
+    def _async_write_ha_state(self) -> None:
```

Comment:
Question for reviewers: This does not seem like a common pattern, so wanted to double check there isn't a better way. Essentially, we want to send an update every type there is a potential state change or refresh of data that works across all the todo entity integrations, given the list of items is updated via state change.

(I added test coverage for all integrations because they all have slightly unique ways they update state)

Reply:
I think it's ok. Platforms or other parts of integrations should use an event listener to track state changes, but if we would do that here, I think it would be unnecessary overhead, performance wise.